File: packages/zynth-core/ios/ZynthKit/src/runtime/ZynthStackSymbolicator.swift
import Foundation

@objc(ZynthStackSymbolicator)
public final class ZynthStackSymbolicator: NSObject {
  private static let logTag = "[ZynthSymbolicator]"
  private struct StackFrame {
    let file: String
    let method: String
    let lineNumber: Int
    let column: Int?
  }

  private struct ParsedLocation {
    let file: String
    let lineNumber: Int?
    let column: Int?
  }

  private struct Segment {
    let generatedColumn: Int
    let sourceIndex: Int?
    let originalLine: Int?
    let originalColumn: Int?
    let nameIndex: Int?
  }

  private struct SourceMapIndex {
    let sources: [String]
    let names: [String]
    let sourceRoot: String?
    let lines: [[Segment]]
  }

  private struct OriginalPosition {
    let source: String?
    let line: Int?
    let column: Int?
    let name: String?
  }

  private static let queue = DispatchQueue(label: "com.zynth.stackSymbolicator")
  private static let lock = NSLock()
  private static var mapCache: [String: SourceMapIndex?] = [:]

  @objc(symbolicate:completion:)
  public static func symbolicate(_ stack: String, completion: @escaping (NSString) -> Void) {
    queue.async {
      let frames = parseStack(stack)
      if frames.isEmpty {
        print("\(logTag) no frames parsed")
        DispatchQueue.main.async { completion(stack as NSString) }
        return
      }
      print("\(logTag) parsed \(frames.count) frames")

      let group = DispatchGroup()
      var outputs = Array(repeating: "", count: frames.count)

      for (index, frame) in frames.enumerated() {
        guard let bundleURL = normalizedBundleURL(for: frame.file) else {
          print("\(logTag) skipping non-symbolic frame: \(frame.file)")
          outputs[index] = formatFrame(method: frame.method, source: frame.file, line: frame.lineNumber, column: frame.column)
          continue
        }

        group.enter()
        symbolicate(frame: frame, bundleURL: bundleURL) { line in
          outputs[index] = line
          group.leave()
        }
      }

      group.notify(queue: queue) {
        let combined = outputs.joined(separator: "\n")
        DispatchQueue.main.async {
          completion(combined as NSString)
        }
      }
    }
  }

  private static func symbolicate(frame: StackFrame, bundleURL: String, completion: @escaping (String) -> Void) {
    fetchMapIndex(bundleURL: bundleURL) { index in
      guard let index else {
        print("\(logTag) no sourcemap for \(bundleURL)")
        completion(formatFrame(method: frame.method, source: frame.file, line: frame.lineNumber, column: frame.column))
        return
      }

      var original = resolveOriginalPosition(index: index, frame: frame)
      if original.source == nil {
        original = findClosestMapping(index: index, frame: frame)
      }

      if let source = original.source {
        let method = original.name ?? frame.method
        print("\(logTag) symbolicated \(frame.file):\(frame.lineNumber)")
        completion(formatFrame(method: method, source: source, line: original.line, column: original.column))
      } else {
        print("\(logTag) mapping not found for \(frame.file):\(frame.lineNumber)")
        completion(formatFrame(method: frame.method, source: frame.file, line: frame.lineNumber, column: frame.column))
      }
    }
  }

  private static func parseStack(_ stack: String) -> [StackFrame] {
    let lines = stack.split(separator: "\n", omittingEmptySubsequences: false)
    var frames: [StackFrame] = []
    for line in lines {
      let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
      guard trimmed.hasPrefix("at ") else { continue }
      let body = String(trimmed.dropFirst(3)).trimmingCharacters(in: .whitespacesAndNewlines)
      if body.isEmpty { continue }

      var method = "<unknown>"
      var location = body

      if let parenStart = body.range(of: " ("),
         body.hasSuffix(")") {
        let m = body[..<parenStart.lowerBound].trimmingCharacters(in: .whitespacesAndNewlines)
        method = m.isEmpty ? "<unknown>" : m
        location = String(body[parenStart.upperBound..<body.index(before: body.endIndex)]).trimmingCharacters(in: .whitespacesAndNewlines)
      }

      let parsed = parseLocation(location)
      guard let lineNumber = parsed.lineNumber else { continue }
      frames.append(StackFrame(file: parsed.file, method: method, lineNumber: lineNumber, column: parsed.column))
    }
    return frames
  }

  private static func parseLocation(_ location: String) -> ParsedLocation {
    var loc = location.trimmingCharacters(in: .whitespacesAndNewlines)
    if loc.hasSuffix(")") {
      loc.removeLast()
    }
    guard let regex = try? NSRegularExpression(pattern: ":(\\d+)(?::(\\d+))?$"),
          let match = regex.firstMatch(in: loc, options: [], range: NSRange(loc.startIndex..., in: loc)),
          let fileRange = Range(NSRange(location: 0, length: match.range.location), in: loc),
          let lineRange = Range(match.range(at: 1), in: loc)
    else {
      return ParsedLocation(file: loc, lineNumber: nil, column: nil)
    }
    let file = String(loc[fileRange])
    let line = Int(loc[lineRange])
    var col: Int?
    if match.numberOfRanges > 2,
       match.range(at: 2).location != NSNotFound,
       let colRange = Range(match.range(at: 2), in: loc) {
      col = Int(loc[colRange])
    }
    return ParsedLocation(file: file, lineNumber: line, column: col)
  }

  private static func fetchMapIndex(bundleURL: String, completion: @escaping (SourceMapIndex?) -> Void) {
    guard let mapURL = mapURLFor(bundleURL: bundleURL) else {
      print("\(logTag) invalid bundle URL: \(bundleURL)")
      completion(nil)
      return
    }
    lock.lock()
    if mapCache.keys.contains(mapURL.absoluteString) {
      let cached = mapCache[mapURL.absoluteString] ?? nil
      lock.unlock()
      completion(cached)
      return
    }
    lock.unlock()

    var request = URLRequest(url: mapURL)
    request.timeoutInterval = 2.0
    request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
    request.setValue("application/json", forHTTPHeaderField: "Accept")

    URLSession.shared.dataTask(with: request) { data, response, _ in
      guard let http = response as? HTTPURLResponse,
            (200...299).contains(http.statusCode),
            let data,
            let json = String(data: data, encoding: .utf8)
      else {
        let status = (response as? HTTPURLResponse)?.statusCode ?? -1
        print("\(logTag) sourcemap fetch failed (\(status)) \(mapURL.absoluteString)")
        lock.lock()
        mapCache[mapURL.absoluteString] = nil
        lock.unlock()
        completion(nil)
        return
      }
      print("\(logTag) sourcemap fetched \(mapURL.absoluteString)")
      let parsed = parseSourceMap(json)
      if parsed == nil {
        print("\(logTag) sourcemap parse failed \(mapURL.absoluteString)")
      }
      lock.lock()
      mapCache[mapURL.absoluteString] = parsed
      lock.unlock()
      completion(parsed)
    }.resume()
  }

  private static func normalizedBundleURL(for file: String) -> String? {
    if file.hasPrefix("http://") || file.hasPrefix("https://") {
      return file
    }
    guard let baseURL = currentDevServerURL() else {
      return nil
    }
    let trimmed = file.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    guard !trimmed.isEmpty else { return nil }
    return baseURL.appendingPathComponent(trimmed).absoluteString
  }

  private static func currentDevServerURL() -> URL? {
    let env = ProcessInfo.processInfo.environment
    if let envValue = env["ZYNTH_DEV_SERVER_URL"]?.trimmingCharacters(in: .whitespacesAndNewlines),
       !envValue.isEmpty,
       let url = URL(string: envValue) {
      return url
    }
    if let plist = Bundle.main.object(forInfoDictionaryKey: "ZynthDevServerURL") as? String {
      let value = plist.trimmingCharacters(in: .whitespacesAndNewlines)
      if !value.isEmpty {
        return URL(string: value)
      }
    }
    return nil
  }

  private static func mapURLFor(bundleURL: String) -> URL? {
    guard var components = URLComponents(string: bundleURL) else { return nil }
    components.path = components.path + ".map"
    return components.url
  }

  private static func parseSourceMap(_ json: String) -> SourceMapIndex? {
    guard let data = json.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let mappings = object["mappings"] as? String,
          !mappings.isEmpty
    else { return nil }

    let sources = (object["sources"] as? [String]) ?? []
    let names = (object["names"] as? [String]) ?? []
    let sourceRoot = (object["sourceRoot"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
    let lines = parseMappings(mappings)
    if lines.isEmpty { return nil }
    return SourceMapIndex(
      sources: sources,
      names: names,
      sourceRoot: sourceRoot?.isEmpty == true ? nil : sourceRoot,
      lines: lines
    )
  }

  private static func parseMappings(_ mappings: String) -> [[Segment]] {
    var lines: [[Segment]] = []
    var previousSource = 0
    var previousOriginalLine = 0
    var previousOriginalColumn = 0
    var previousName = 0

    for lineChunk in mappings.split(separator: ";", omittingEmptySubsequences: false) {
      var previousGeneratedColumn = 0
      var segments: [Segment] = []
      for raw in lineChunk.split(separator: ",", omittingEmptySubsequences: true) {
        let values = decodeVlq(String(raw))
        if values.isEmpty { continue }
        let generatedColumn = previousGeneratedColumn + values[0]
        previousGeneratedColumn = generatedColumn
        if values.count >= 4 {
          let sourceIndex = previousSource + values[1]
          previousSource = sourceIndex
          let originalLine = previousOriginalLine + values[2]
          previousOriginalLine = originalLine
          let originalColumn = previousOriginalColumn + values[3]
          previousOriginalColumn = originalColumn
          var nameIndex: Int?
          if values.count >= 5 {
            nameIndex = previousName + values[4]
            previousName = nameIndex ?? previousName
          }
          segments.append(
            Segment(
              generatedColumn: generatedColumn,
              sourceIndex: sourceIndex,
              originalLine: originalLine,
              originalColumn: originalColumn,
              nameIndex: nameIndex
            )
          )
        } else {
          segments.append(
            Segment(
              generatedColumn: generatedColumn,
              sourceIndex: nil,
              originalLine: nil,
              originalColumn: nil,
              nameIndex: nil
            )
          )
        }
      }
      lines.append(segments)
    }
    return lines
  }

  private static func decodeVlq(_ segment: String) -> [Int] {
    var values: [Int] = []
    var chars = Array(segment)
    var index = 0
    while index < chars.count {
      var shift = 0
      var result = 0
      var continuation = false
      repeat {
        if index >= chars.count { return values }
        let digit = fromBase64(chars[index])
        if digit < 0 { return values }
        index += 1
        continuation = (digit & 32) != 0
        let valueBits = digit & 31
        result += valueBits << shift
        shift += 5
      } while continuation
      let isNegative = (result & 1) == 1
      let decoded = result >> 1
      values.append(isNegative ? -decoded : decoded)
    }
    return values
  }

  private static func fromBase64(_ c: Character) -> Int {
    guard let scalar = c.unicodeScalars.first?.value else { return -1 }
    switch scalar {
    case 65...90: return Int(scalar - 65)
    case 97...122: return 26 + Int(scalar - 97)
    case 48...57: return 52 + Int(scalar - 48)
    case 43: return 62
    case 47: return 63
    default: return -1
    }
  }

  private static func resolveOriginalPosition(index: SourceMapIndex, frame: StackFrame) -> OriginalPosition {
    let line = frame.lineNumber
    let colInput = frame.column ?? 1
    let colZero = max(0, colInput - 1)
    let candidates = [
      lookupPosition(index: index, generatedLine: line, generatedColumn: colZero, greatestLowerBound: true),
      lookupPosition(index: index, generatedLine: line, generatedColumn: colInput, greatestLowerBound: true),
      lookupPosition(index: index, generatedLine: line, generatedColumn: colZero, greatestLowerBound: false),
      lookupPosition(index: index, generatedLine: line, generatedColumn: colInput, greatestLowerBound: false),
    ]
    for item in candidates where item.source != nil {
      return item
    }
    return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
  }

  private static func findClosestMapping(index: SourceMapIndex, frame: StackFrame) -> OriginalPosition {
    guard frame.lineNumber - 1 >= 0, frame.lineNumber - 1 < index.lines.count else {
      return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
    }
    let segments = index.lines[frame.lineNumber - 1]
    if segments.isEmpty {
      return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
    }
    let targetColumn = frame.column ?? 0
    var bestSegment: Segment?
    var bestDelta = Int.max
    for segment in segments {
      guard let sourceIndex = segment.sourceIndex,
            sourceIndex >= 0,
            sourceIndex < index.sources.count
      else { continue }
      let delta = abs(segment.generatedColumn - targetColumn)
      if delta < bestDelta {
        bestDelta = delta
        bestSegment = segment
      }
    }
    return segmentToOriginal(index: index, segment: bestSegment)
  }

  private static func lookupPosition(
    index: SourceMapIndex,
    generatedLine: Int,
    generatedColumn: Int,
    greatestLowerBound: Bool
  ) -> OriginalPosition {
    guard generatedLine - 1 >= 0, generatedLine - 1 < index.lines.count else {
      return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
    }
    let segments = index.lines[generatedLine - 1]
    if segments.isEmpty {
      return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
    }
    let chosen: Segment?
    if greatestLowerBound {
      chosen = segments.last(where: { $0.generatedColumn <= generatedColumn })
    } else {
      chosen = segments.first(where: { $0.generatedColumn >= generatedColumn })
    }
    return segmentToOriginal(index: index, segment: chosen)
  }

  private static func segmentToOriginal(index: SourceMapIndex, segment: Segment?) -> OriginalPosition {
    guard let segment,
          let sourceIndex = segment.sourceIndex,
          sourceIndex >= 0,
          sourceIndex < index.sources.count
    else {
      return OriginalPosition(source: nil, line: nil, column: nil, name: nil)
    }
    let source = resolveSource(sourceRoot: index.sourceRoot, source: index.sources[sourceIndex])
    let name: String?
    if let nameIndex = segment.nameIndex,
       nameIndex >= 0,
       nameIndex < index.names.count {
      name = index.names[nameIndex]
    } else {
      name = nil
    }
    return OriginalPosition(
      source: source,
      line: segment.originalLine.map { $0 + 1 },
      column: segment.originalColumn,
      name: name
    )
  }

  private static func resolveSource(sourceRoot: String?, source: String) -> String {
    guard let sourceRoot, !sourceRoot.isEmpty else { return source }
    if source.hasPrefix("/") || source.contains("://") { return source }
    return sourceRoot.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/" + source.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
  }

  private static func formatFrame(method: String, source: String?, line: Int?, column: Int?) -> String {
    let safeSource = source ?? "<unknown>"
    let safeLine = line.map(String.init) ?? "null"
    let safeColumn = column.map(String.init) ?? "null"
    return "at \(method) (\(safeSource):\(safeLine):\(safeColumn))"
  }
}

File: packages/zynth-core/android/ZynthKit/src/main/java/com/zynth/kit/runtime/ZynthStackSymbolicator.kt
package com.zynth.kit.runtime

import android.os.Handler
import android.os.Looper
import android.util.Log
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import org.json.JSONArray
import org.json.JSONObject

internal object ZynthStackSymbolicator {
  private const val TAG = "ZynthSymbolicator"
  private val executor = Executors.newSingleThreadExecutor { runnable ->
    Thread(runnable, "ZynthStackSymbolicator").apply { isDaemon = true }
  }
  private val mainHandler = Handler(Looper.getMainLooper())
  private val mapCache = ConcurrentHashMap<String, SourceMapIndex?>()

  private data class StackFrame(
    val file: String,
    val method: String,
    val lineNumber: Int,
    val column: Int?,
  )

  private data class ParsedLocation(
    val file: String,
    val lineNumber: Int?,
    val column: Int?,
  )

  private data class OriginalPosition(
    val source: String?,
    val line: Int?,
    val column: Int?,
    val name: String?,
  )

  private data class Segment(
    val generatedColumn: Int,
    val sourceIndex: Int?,
    val originalLine: Int?,
    val originalColumn: Int?,
    val nameIndex: Int?,
  )

  private data class SourceMapIndex(
    val sources: List<String>,
    val names: List<String>,
    val sourceRoot: String?,
    val lines: List<List<Segment>>,
  )

  fun symbolicateStackTrace(stack: String, callback: (String) -> Unit) {
    executor.execute {
      val symbolicated = runCatching { symbolicateSync(stack) }.getOrElse { stack }
      mainHandler.post { callback(symbolicated) }
    }
  }

  private fun symbolicateSync(stack: String): String {
    val frames = parseStack(stack)
    if (frames.isEmpty()) {
      Log.d(TAG, "no frames parsed")
      return stack
    }
    Log.d(TAG, "parsed ${frames.size} frames")
    val symbolicated = ArrayList<String>(frames.size)

    for (frame in frames) {
      val bundleUrl = normalizedBundleUrl(frame.file)
      if (bundleUrl == null) {
        Log.d(TAG, "skipping non-symbolic frame: ${frame.file}")
        symbolicated += formatFrame(frame.method, frame.file, frame.lineNumber, frame.column)
        continue
      }

      val index = fetchMapIndex(bundleUrl)
      if (index == null) {
        Log.d(TAG, "no sourcemap for $bundleUrl")
        symbolicated += formatFrame(frame.method, frame.file, frame.lineNumber, frame.column)
        continue
      }

      var original = resolveOriginalPosition(index, frame)
      if (original.source == null) {
        original = findClosestMapping(index, frame)
      }

      if (original.source != null) {
        val method = original.name ?: frame.method
        Log.d(TAG, "symbolicated ${frame.file}:${frame.lineNumber}")
        symbolicated += formatFrame(method, original.source, original.line, original.column)
      } else {
        Log.d(TAG, "mapping not found for ${frame.file}:${frame.lineNumber}")
        symbolicated += formatFrame(frame.method, frame.file, frame.lineNumber, frame.column)
      }
    }

    return symbolicated.joinToString("\n")
  }

  private fun parseStack(stack: String): List<StackFrame> {
    val lines = stack.split('\n')
    val frames = mutableListOf<StackFrame>()
    for (line in lines) {
      val trimmed = line.trim()
      if (!trimmed.startsWith("at ")) continue
      val body = trimmed.removePrefix("at ").trim()
      if (body.isEmpty()) continue

      var method = "<unknown>"
      var location = body

      val parenStart = body.indexOf(" (")
      if (parenStart != -1 && body.endsWith(")")) {
        method = body.substring(0, parenStart).trim().ifBlank { "<unknown>" }
        location = body.substring(parenStart + 2, body.length - 1).trim()
      }

      val parsed = parseLocation(location)
      if (parsed.lineNumber == null) continue
      frames += StackFrame(
        file = parsed.file,
        method = method,
        lineNumber = parsed.lineNumber,
        column = parsed.column,
      )
    }
    return frames
  }

  private fun parseLocation(location: String): ParsedLocation {
    var loc = location.trim()
    if (loc.endsWith(")")) {
      loc = loc.substring(0, loc.length - 1)
    }
    val match = Regex(":(\\d+)(?::(\\d+))?$").find(loc)
      ?: return ParsedLocation(file = loc, lineNumber = null, column = null)
    val lineNumber = match.groupValues.getOrNull(1)?.toIntOrNull()
    val column = match.groupValues.getOrNull(2)?.toIntOrNull()
    val file = loc.substring(0, match.range.first)
    return ParsedLocation(file = file, lineNumber = lineNumber, column = column)
  }

  private fun fetchMapIndex(bundleUrl: String): SourceMapIndex? {
    val mapUrl = mapUrlFor(bundleUrl) ?: return null
    if (mapCache.containsKey(mapUrl)) {
      return mapCache[mapUrl]
    }
    val loaded = runCatching {
      val mapJson = fetchMapJson(mapUrl) ?: return@runCatching null
      parseSourceMap(mapJson)
    }.getOrNull()
    if (loaded == null) {
      Log.d(TAG, "sourcemap parse failed $mapUrl")
    }
    mapCache[mapUrl] = loaded
    return loaded
  }

  private fun normalizedBundleUrl(file: String): String? {
    if (file.startsWith("http://") || file.startsWith("https://")) {
      return file
    }
    val baseUrl = System.getProperty("ZYNTH_DEV_SERVER_URL")?.trim().orEmpty()
    if (baseUrl.isEmpty()) {
      return null
    }
    val cleaned = file.trim().trimStart('/')
    if (cleaned.isEmpty()) {
      return null
    }
    return runCatching { URL(URL(baseUrl), cleaned).toString() }.getOrNull()
  }

  private fun mapUrlFor(bundleUrl: String): String? {
    return try {
      val uri = URI(bundleUrl)
      val path = uri.path ?: return null
      URI(
        uri.scheme,
        uri.userInfo,
        uri.host,
        uri.port,
        "$path.map",
        uri.query,
        uri.fragment,
      ).toString()
    } catch (_: Throwable) {
      null
    }
  }

  private fun fetchMapJson(mapUrl: String): String? {
    val connection = (URI(mapUrl).toURL().openConnection() as? HttpURLConnection) ?: return null
    return try {
      connection.requestMethod = "GET"
      connection.connectTimeout = 1500
      connection.readTimeout = 2000
      connection.setRequestProperty("Accept", "application/json")
      val status = connection.responseCode
      if (status !in 200..299) {
        Log.d(TAG, "sourcemap fetch failed ($status) $mapUrl")
        return null
      }
      Log.d(TAG, "sourcemap fetched $mapUrl")
      BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
        buildString {
          var line: String? = reader.readLine()
          while (line != null) {
            append(line)
            line = reader.readLine()
          }
        }
      }
    } catch (_: Throwable) {
      null
    } finally {
      connection.disconnect()
    }
  }

  private fun parseSourceMap(json: String): SourceMapIndex? {
    val root = runCatching { JSONObject(json) }.getOrNull() ?: return null
    val mappings = root.optString("mappings", "")
    if (mappings.isBlank()) return null
    val sources = jsonArrayToStringList(root.optJSONArray("sources"))
    val names = jsonArrayToStringList(root.optJSONArray("names"))
    val sourceRoot = root.optString("sourceRoot", "").ifBlank { null }
    val lines = parseMappings(mappings)
    if (lines.isEmpty()) return null
    return SourceMapIndex(
      sources = sources,
      names = names,
      sourceRoot = sourceRoot,
      lines = lines,
    )
  }

  private fun jsonArrayToStringList(array: JSONArray?): List<String> {
    if (array == null) return emptyList()
    val out = ArrayList<String>(array.length())
    for (i in 0 until array.length()) {
      out += array.optString(i, "")
    }
    return out
  }

  private fun parseMappings(mappings: String): List<List<Segment>> {
    val lines = ArrayList<MutableList<Segment>>()
    var previousSource = 0
    var previousOriginalLine = 0
    var previousOriginalColumn = 0
    var previousName = 0

    val lineChunks = mappings.split(';')
    for (lineChunk in lineChunks) {
      var previousGeneratedColumn = 0
      val segments = mutableListOf<Segment>()
      if (lineChunk.isNotEmpty()) {
        val rawSegments = lineChunk.split(',')
        for (raw in rawSegments) {
          if (raw.isEmpty()) continue
          val values = decodeVlq(raw)
          if (values.isEmpty()) continue
          val generatedColumn = previousGeneratedColumn + values[0]
          previousGeneratedColumn = generatedColumn
          if (values.size >= 4) {
            val sourceIndex = previousSource + values[1]
            previousSource = sourceIndex
            val originalLine = previousOriginalLine + values[2]
            previousOriginalLine = originalLine
            val originalColumn = previousOriginalColumn + values[3]
            previousOriginalColumn = originalColumn
            var nameIndex: Int? = null
            if (values.size >= 5) {
              nameIndex = previousName + values[4]
              previousName = nameIndex
            }
            segments += Segment(
              generatedColumn = generatedColumn,
              sourceIndex = sourceIndex,
              originalLine = originalLine,
              originalColumn = originalColumn,
              nameIndex = nameIndex,
            )
          } else {
            segments += Segment(
              generatedColumn = generatedColumn,
              sourceIndex = null,
              originalLine = null,
              originalColumn = null,
              nameIndex = null,
            )
          }
        }
      }
      lines += segments
    }
    return lines
  }

  private fun decodeVlq(segment: String): List<Int> {
    val values = mutableListOf<Int>()
    var index = 0
    while (index < segment.length) {
      var shift = 0
      var result = 0
      var continuation: Boolean
      do {
        if (index >= segment.length) return values
        val digit = fromBase64(segment[index])
        if (digit < 0) return values
        index += 1
        continuation = (digit and 32) != 0
        val valueBits = digit and 31
        result += valueBits shl shift
        shift += 5
      } while (continuation)
      val isNegative = (result and 1) == 1
      val decoded = result shr 1
      values += if (isNegative) -decoded else decoded
    }
    return values
  }

  private fun fromBase64(c: Char): Int {
    return when (c) {
      in 'A'..'Z' -> c.code - 'A'.code
      in 'a'..'z' -> 26 + c.code - 'a'.code
      in '0'..'9' -> 52 + c.code - '0'.code
      '+' -> 62
      '/' -> 63
      else -> -1
    }
  }

  private fun resolveOriginalPosition(index: SourceMapIndex, frame: StackFrame): OriginalPosition {
    val line = frame.lineNumber
    val colInput = frame.column ?: 1
    val colZero = maxOf(0, colInput - 1)

    val candidates = listOf(
      lookupPosition(index, line, colZero, greatestLowerBound = true),
      lookupPosition(index, line, colInput, greatestLowerBound = true),
      lookupPosition(index, line, colZero, greatestLowerBound = false),
      lookupPosition(index, line, colInput, greatestLowerBound = false),
    )
    for (candidate in candidates) {
      if (candidate.source != null) return candidate
    }
    return OriginalPosition(source = null, line = null, column = null, name = null)
  }

  private fun findClosestMapping(index: SourceMapIndex, frame: StackFrame): OriginalPosition {
    val segments = index.lines.getOrNull(frame.lineNumber - 1).orEmpty()
    if (segments.isEmpty()) return OriginalPosition(source = null, line = null, column = null, name = null)
    val targetColumn = frame.column ?: 0
    var best: Segment? = null
    var bestDelta = Int.MAX_VALUE
    for (segment in segments) {
      val sourceIndex = segment.sourceIndex ?: continue
      if (sourceIndex < 0 || sourceIndex >= index.sources.size) continue
      val delta = kotlin.math.abs(segment.generatedColumn - targetColumn)
      if (delta < bestDelta) {
        bestDelta = delta
        best = segment
      }
    }
    return segmentToOriginal(index, best)
  }

  private fun lookupPosition(
    index: SourceMapIndex,
    generatedLine: Int,
    generatedColumn: Int,
    greatestLowerBound: Boolean,
  ): OriginalPosition {
    val segments = index.lines.getOrNull(generatedLine - 1).orEmpty()
    if (segments.isEmpty()) return OriginalPosition(source = null, line = null, column = null, name = null)
    val chosen = if (greatestLowerBound) {
      var best: Segment? = null
      for (segment in segments) {
        if (segment.generatedColumn <= generatedColumn) {
          best = segment
        } else {
          break
        }
      }
      best
    } else {
      var best: Segment? = null
      for (segment in segments) {
        if (segment.generatedColumn >= generatedColumn) {
          best = segment
          break
        }
      }
      best
    }
    return segmentToOriginal(index, chosen)
  }

  private fun segmentToOriginal(index: SourceMapIndex, segment: Segment?): OriginalPosition {
    val sourceIndex = segment?.sourceIndex ?: return OriginalPosition(null, null, null, null)
    if (sourceIndex < 0 || sourceIndex >= index.sources.size) {
      return OriginalPosition(null, null, null, null)
    }
    val source = resolveSource(index.sourceRoot, index.sources[sourceIndex])
    val name = segment.nameIndex?.let { idx ->
      if (idx in 0 until index.names.size) index.names[idx] else null
    }
    return OriginalPosition(
      source = source,
      line = segment.originalLine?.plus(1),
      column = segment.originalColumn,
      name = name,
    )
  }

  private fun resolveSource(sourceRoot: String?, source: String): String {
    if (sourceRoot.isNullOrBlank() || source.startsWith("/") || source.contains("://")) {
      return source
    }
    val left = sourceRoot.trimEnd('/')
    val right = source.trimStart('/')
    return "$left/$right"
  }

  private fun formatFrame(method: String, source: String?, line: Int?, column: Int?): String {
    val safeSource = source ?: "<unknown>"
    val safeLine = line?.toString() ?: "null"
    val safeColumn = column?.toString() ?: "null"
    return "at $method ($safeSource:$safeLine:$safeColumn)"
  }
}
